In [3]:
from sklearn.feature_extraction.text import CountVectorizer
import json
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [4]:
with open('pizza_request_dataset.json') as f:
  data = json.load(f)

In [5]:
df = pd.read_json('pizza_request_dataset.json')
train, test = train_test_split(df, test_size=567, random_state=42)

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
word_vectorizer = CountVectorizer(ngram_range=(1,1), analyzer='word')
sparse_matrix = word_vectorizer.fit_transform(train['request_text'])
frequencies = sum(sparse_matrix).toarray()[0]
train_ngrams = pd.DataFrame(frequencies, index=word_vectorizer.get_feature_names(), columns=['frequency'])

In [ ]:
topp_500_unigrams = train_ngrams.sort_values("frequency", ascending=False).head(500)

In [43]:
from sklearn.feature_extraction.text import CountVectorizer
word_vectorizer = CountVectorizer(ngram_range=(2,2), analyzer='word')
sparse_matrix = word_vectorizer.fit_transform(train['request_text'])
frequencies = sum(sparse_matrix).toarray()[0]
train_bigrams = pd.DataFrame(frequencies, index=word_vectorizer.get_feature_names(), columns=['frequency'])

In [44]:
topp500_bigrams = train_bigrams.sort_values("frequency", ascending=False).head(500)

In [45]:
topp500_bigrams['bigrams'] = topp500_bigrams.index

In [46]:
topp_500_unigrams['unigrams'] = topp_500_unigrams.index
unigrams = topp_500_unigrams['unigrams'].tolist()
bigrams =topp500_bigrams['bigrams'].tolist()
unigrams
ngrams = unigrams + bigrams
#ngrams

In [47]:
#create X to be 5104x1000 shape
X = []
for text in train['request_text'].tolist():
  contains_ngram =[]
  for i in range(len(ngrams)):
    if ngrams[i] in text:
      contains_ngram.append(1)
    else:
      contains_ngram.append(0)
  X.append(contains_ngram)

In [48]:
X = np.array(X)
X.shape

(5104, 1000)

In [6]:
y = train['requester_received_pizza']

In [7]:
y_int = [int(val == True) for val in y]
len(y_int)

5104

In [51]:
from sklearn import svm
from sklearn import linear_model
clf = svm.SVC(kernel='linear')
clf.fit(X, y_int)
#svm = sklearn.linear_model.SGDClassifier(loss='hinge')

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [52]:
#sklearn.linear_model.SGDClassifier¶X_test = []
X_test = []
for text in test['request_text'].tolist():
  contains_ngram =[]
  for i in range(len(ngrams)):
    if ngrams[i] in text:
      contains_ngram.append(1)
    else:
      contains_ngram.append(0)
  X_test.append(contains_ngram)
y_test = test['requester_received_pizza']
y_test = [int(val == True) for val in y_test]


In [53]:
test_set_preds = clf.predict(X_test)

In [54]:
import sklearn.metrics

accuracy_1a = clf.score(X_test, y_test)
precision_1a = sklearn.metrics.precision_score(y_test, test_set_preds)
recall_1a = sklearn.metrics.recall_score(y_test, test_set_preds)
f1_1a = sklearn.metrics.f1_score(y_test, test_set_preds)
auc_1a= sklearn.metrics.roc_auc_score(y_test, test_set_preds)

tn, fp, fn, tp = sklearn.metrics.confusion_matrix(y_test, test_set_preds).ravel()
specificity_1a = tn / (tn+fp)

print("Answers for 1a: ")
print("accuracy: " + str(accuracy_1a))
print("precision: " +str(precision_1a))
print("recall: " + str(recall_1a))
print("f1: " + str(f1_1a))
print("auc: " + str(auc_1a))
print("specificity: " + str(specificity_1a))

Answers for 1a: 
accuracy: 0.7072310405643739
precision: 0.3924050632911392
recall: 0.2080536912751678
f1: 0.2719298245614035
auc: 0.5466105776950002
specificity: 0.8851674641148325


Problem 1b

In [12]:
feature_cols = ['post_was_edited','requester_account_age_in_days_at_request', 'requester_account_age_in_days_at_retrieval', #,
                'requester_days_since_first_post_on_raop_at_request','requester_days_since_first_post_on_raop_at_retrieval',
                'requester_number_of_comments_at_request','requester_number_of_comments_at_retrieval','requester_number_of_comments_in_raop_at_request',
                'requester_number_of_comments_in_raop_at_retrieval','requester_number_of_posts_at_request','requester_number_of_posts_at_retrieval',
                'requester_number_of_posts_on_raop_at_request','requester_number_of_posts_on_raop_at_retrieval','requester_number_of_subreddits_at_request',
                'requester_subreddits_at_request', 'number_of_downvotes_of_request_at_retrieval', 'number_of_upvotes_of_request_at_retrieval',
                'requester_upvotes_minus_downvotes_at_request', 'requester_upvotes_minus_downvotes_at_retrieval', 'requester_upvotes_plus_downvotes_at_request',
                'requester_upvotes_plus_downvotes_at_retrieval', 'requester_user_flair']
              

In [13]:
X_train_1b = train.loc[:, feature_cols]
X_test_1b = test.loc[:, feature_cols]

In [14]:
subreddits = X_train_1b['requester_subreddits_at_request']
from itertools import chain, groupby
from collections import Counter
ll = subreddits.tolist()
top_subs = sorted(chain(*(Counter(l).items() for l in ll)), reverse=True)

In [15]:
Counter(chain.from_iterable(ll)).most_common(10)

[('AskReddit', 2904),
 ('pics', 2454),
 ('funny', 2412),
 ('IAmA', 1914),
 ('WTF', 1913),
 ('gaming', 1867),
 ('Random_Acts_Of_Pizza', 1759),
 ('videos', 1455),
 ('todayilearned', 1400),
 ('AdviceAnimals', 1308)]

Most Common Subreddits : ('AskReddit', 2904),
 ('pics', 2454),
 ('funny', 2412),
 ('IAmA', 1914),
 ('WTF', 1913),
 ('gaming', 1867),
 ('Random_Acts_Of_Pizza', 1759),
 ('videos', 1455),
 ('todayilearned', 1400),
 ('AdviceAnimals', 1308)]

In [ ]:
most_common_subs = ['AskReddit', 'pics', 'funny', 'IAmA', "WTF", 'gaming', 'Random_Acts_Of_Pizza', 'videos', 'todayilearned', 'AdviceAnimals']
sub_reddit_feats_train = []
for text in train['requester_subreddits_at_request'].tolist():
  print(text)
  contains_sub =[]
  for i in range(len(most_common_subs)):
    if most_common_subs[i] in text:
      contains_sub.append(1)
    else:
      contains_sub.append(0)
  sub_reddit_feats_train.append(contains_sub)
sub_reddit_feats_train

In [17]:
np.array(sub_reddit_feats_train).sum()

19386

In [18]:
np_subreddits_train = np.array(sub_reddit_feats_train)
np_subreddits_train.shape

(5104, 10)

In [19]:
most_common_subs = ['AskReddit', 'pics', 'funny', 'IAmA', "WTF", 'gaming', 'Random_Acts_Of_Pizza', 'videos', 'todayilearned', 'AdviceAnimals']
sub_reddit_feats_test = []
for text in test['requester_subreddits_at_request'].tolist():
  contains_sub =[]
  for i in range(len(most_common_subs)):
    if most_common_subs[i] in text:
      contains_sub.append(1)
    else:
      contains_sub.append(0)
  sub_reddit_feats_test.append(contains_sub)
np_subreddits_test = np.array(sub_reddit_feats_test)
np_subreddits_test.shape

(567, 10)

In [20]:
np_subreddits_test

array([[0, 0, 0, ..., 0, 0, 0],
       [1, 1, 0, ..., 0, 0, 0],
       [1, 1, 1, ..., 1, 1, 1],
       ...,
       [1, 1, 1, ..., 1, 1, 1],
       [1, 1, 1, ..., 1, 1, 0],
       [1, 1, 0, ..., 0, 0, 1]])

In [21]:
#X_train_1b.drop(['F','I','P','h','m','o','r','s'], axis=1, inplace=True)
one_hot_flair = pd.get_dummies(X_train_1b['requester_user_flair'])
#one_hot_flair

In [22]:
X_train_1b_final = pd.concat([X_train_1b,one_hot_flair], axis=1)
X_train_1b_final = X_train_1b_final.drop(['requester_user_flair','requester_subreddits_at_request'],1)
#X_train_1b_final



In [69]:
X_train_1b_final

post_was_edited  requester_account_age_in_days_at_request  ...  PIF  shroom
4023                0                                 23.688519  ...    0       0
2988                0                                511.161817  ...    0       0
3313                1                                 60.862581  ...    0       0
3409                0                                 82.156586  ...    0       0
676                 0                                387.013611  ...    0       0
...               ...                                       ...  ...  ...     ...
3772                0                                542.096296  ...    0       0
5191                0                                732.137812  ...    0       0
5226                0                                239.057743  ...    0       0
5390                0                                 72.029942  ...    0       0
860                 0                                135.643889  ...    0       0

[5104 rows x 22 columns]

In [23]:
one_hot_flair = pd.get_dummies(X_test_1b['requester_user_flair'])
X_test_1b_final = pd.concat([X_test_1b,one_hot_flair], axis=1)
X_test_1b_final = X_test_1b_final.drop(['requester_user_flair','requester_subreddits_at_request'],1)
X_test_1b_final

post_was_edited  requester_account_age_in_days_at_request  ...  PIF  shroom
3671                0                                  0.000000  ...    0       0
5183                0                                205.548843  ...    0       0
1859                0                                411.171655  ...    0       1
2589                0                                105.919167  ...    0       0
1662                0                                362.890451  ...    0       1
...               ...                                       ...  ...  ...     ...
3817                0                                  0.848299  ...    0       0
151                 0                                 11.471829  ...    0       1
763                 0                                500.034468  ...    0       0
3724                0                               1297.815868  ...    0       0
2464                0                                202.401644  ...    0       0

[567 rows x 22 columns]

In [25]:
columns_not_to_normalize = ['post_was_edited', 'PIF', 'shroom', ]

In [81]:
columns =X_test_1b_final.columns.tolist()

In [86]:
len(columns)

22

In [26]:
def normalize(dataset, list_columns_keep):
    dataNorm=((dataset-dataset.min())/(dataset.max()-dataset.min()))
    for col in list_columns_keep:
      dataNorm[col]=dataset[col]
    return dataNorm

In [27]:
X_train_norm = normalize(X_train_1b_final,columns_not_to_normalize  )
X_test_norm = normalize(X_test_1b_final, columns_not_to_normalize)
X_train_norm

post_was_edited  requester_account_age_in_days_at_request  ...  PIF  shroom
4023                0                                  0.008431  ...    0       0
2988                0                                  0.181924  ...    0       0
3313                1                                  0.021661  ...    0       0
3409                0                                  0.029240  ...    0       0
676                 0                                  0.137739  ...    0       0
...               ...                                       ...  ...  ...     ...
3772                0                                  0.192934  ...    0       0
5191                0                                  0.260570  ...    0       0
5226                0                                  0.085081  ...    0       0
5390                0                                  0.025636  ...    0       0
860                 0                                  0.048276  ...    0       0

[5104 rows x 22 columns]

In [28]:
X_test_norm

post_was_edited  requester_account_age_in_days_at_request  ...  PIF  shroom
3671                0                                  0.000000  ...    0       0
5183                0                                  0.091958  ...    0       0
1859                0                                  0.183949  ...    0       1
2589                0                                  0.047386  ...    0       0
1662                0                                  0.162349  ...    0       1
...               ...                                       ...  ...  ...     ...
3817                0                                  0.000380  ...    0       0
151                 0                                  0.005132  ...    0       1
763                 0                                  0.223704  ...    0       0
3724                0                                  0.580614  ...    0       0
2464                0                                  0.090550  ...    0       0

[567 rows x 22 columns]

In [29]:
X_train_np = X_train_norm.to_numpy()
X_test_np = X_test_norm.to_numpy()
X_train_np.shape

(5104, 22)

In [30]:
X_test_np.shape

(567, 22)

In [31]:
final_X_train_np = np.concatenate((X_train_np,np_subreddits_train),axis=1)
final_X_test_np = np.concatenate((X_test_np,np_subreddits_test),axis=1)
final_X_train_np

array([[0.        , 0.00843083, 0.23702223, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.18192425, 0.19772573, ..., 1.        , 0.        ,
        1.        ],
       [1.        , 0.0216612 , 0.26001422, ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.08508148, 0.35723721, ..., 1.        , 1.        ,
        0.        ],
       [0.        , 0.0256357 , 0.08619742, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.04827613, 0.22009133, ..., 0.        , 0.        ,
        0.        ]])

In [34]:
final_X_train_np.shape

(5104, 32)

In [32]:
#normalize
final_X_test_np

array([[0.        , 0.        , 0.07677809, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.09195805, 0.16311123, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.18394918, 0.21120938, ..., 1.        , 1.        ,
        1.        ],
       ...,
       [0.        , 0.22370446, 0.19183984, ..., 1.        , 1.        ,
        1.        ],
       [0.        , 0.58061437, 0.75337477, ..., 1.        , 1.        ,
        0.        ],
       [0.        , 0.09055006, 0.20301815, ..., 0.        , 0.        ,
        1.        ]])

In [35]:
final_X_test_np.shape

(567, 32)

In [ ]:
# clf = svm.SVC(kernel='linear')
# clf.fit(final_X_train_np, y_int)

In [37]:
from sklearn.linear_model import SGDClassifier

In [38]:
clf = SGDClassifier(loss="hinge").fit(final_X_train_np, y_int)

In [39]:
#combine these three dataframes: X_test_1b_no_flair, sub_one_hot_test_df, test_one_hot_flair, 
test_set_preds = clf.predict(final_X_test_np)

In [42]:
y_test = test['requester_received_pizza']
y_test = [int(val == True) for val in y_test]


In [43]:
import sklearn.metrics

accuracy_1a = clf.score(final_X_test_np, y_test)
precision_1a = sklearn.metrics.precision_score(y_test, test_set_preds)
recall_1a = sklearn.metrics.recall_score(y_test, test_set_preds)
f1_1a = sklearn.metrics.f1_score(y_test, test_set_preds)
auc_1a= sklearn.metrics.roc_auc_score(y_test, test_set_preds)

tn, fp, fn, tp = sklearn.metrics.confusion_matrix(y_test, test_set_preds).ravel()
specificity_1a = tn / (tn+fp)


In [ ]:
#clf = SGDClassifier(loss="hinge").fit(X_train_1b_lst, y_int)
import sklearn.metrics

accuracy_1a = clf.score(normed_matrix_test, y_test)
precision_1a = sklearn.metrics.precision_score(y_test, test_set_preds)
recall_1a = sklearn.metrics.recall_score(y_test, test_set_preds)
f1_1a = sklearn.metrics.f1_score(y_test, test_set_preds)
auc_1a= sklearn.metrics.roc_auc_score(y_test, test_set_preds)

tn, fp, fn, tp = sklearn.metrics.confusion_matrix(y_test, test_set_preds).ravel()
specificity_1a = tn / (tn+fp)

In [44]:
print("Answers for 1b: ")
print("accuracy: " + str(accuracy_1a))
print("precision: " +str(precision_1a))
print("recall: " + str(recall_1a))
print("f1: " + str(f1_1a))
print("auc: " + str(auc_1a))
print("specificity " + str(specificity_1a))

Answers for 1b: 
accuracy: 0.9153439153439153
precision: 0.9391304347826087
recall: 0.7248322147651006
f1: 0.8181818181818182
auc: 0.8540429016409234
specificity 0.9832535885167464


In [ ]:
y_test = test['requester_received_pizza']
y_test = [int(val == True) for val in y_test]

Model 3

In [ ]:
desire_set  = set(line.strip() for line in open('desire.txt'))
family_set = set(line.strip() for line in open('family.txt'))
job_set = set(line.strip() for line in open('job.txt'))
money_set = set(line.strip() for line in open('money.txt'))
student_set = set(line.strip() for line in open('student.txt'))

In [ ]:
all_sets = [desire_set, family_set, job_set, money_set, student_set]

In [ ]:
X_train_1c = []
for text in train['request_text'].tolist():
  total_word_count = len(text.split())

  row =[]
  for i in range(len(all_sets)):
    curr_set = all_sets[i]
    set_word_count = 0
    for word in curr_set:
      if word in text:
        set_word_count +=1
    if total_word_count > 0:
      row.append(set_word_count/total_word_count)
    else:
      row.append(0)
  X_train_1c.append(row)

X_train_1c

In [ ]:
X_train_1c_np = np.array(X_train_1c)
X_train_1c_np.shape

(5104, 5)

In [ ]:
X_test_1c = []
for text in test['request_text'].tolist():
  total_word_count = len(text.split())

  row =[]
  for i in range(len(all_sets)):
    curr_set = all_sets[i]
    set_word_count = 0
    for word in curr_set:
      if word in text:
        set_word_count +=1
    if total_word_count > 0:
      row.append(set_word_count/total_word_count)
    else:
      row.append(0)
  X_test_1c.append(row)

X_test_1c
X_test_1c_np = np.array(X_test_1c)
X_test_1c_np.shape

(567, 5)

In [ ]:
y_tr = train['requester_received_pizza']
y_train = [int(val == True) for val in y_tr]

y_te = test['requester_received_pizza']
y_test = [int(val == True) for val in y_te]


In [ ]:
from sklearn import svm
from sklearn import linear_model


In [ ]:
y_train

In [ ]:
#from sklearn import svm
#from sklearn import linear_model

svm3 = linear_model.SGDClassifier(loss='hinge', penalty='l2', alpha=.001).fit(X_train_1c_np, y_train)
#clf = sklearn.linear_model.SGDClassifier(loss="hinge").fit(X_1c, y_train)

In [ ]:
import sklearn.metrics


test_set_preds = svm3.predict(X_test_1c_np)

accuracy_1a = svm3.score(X_test_1c, y_test)
precision_1a = sklearn.metrics.precision_score(y_test, test_set_preds)
recall_1a = sklearn.metrics.recall_score(y_test, test_set_preds)
f1_1a = sklearn.metrics.f1_score(y_test, test_set_preds)
auc_1a= sklearn.metrics.roc_auc_score(y_test, test_set_preds)

tn, fp, fn, tp = sklearn.metrics.confusion_matrix(y_test, test_set_preds).ravel()
specificity_1a = tn / (tn+fp)

print("Answers for 1c: ")
print("accuracy: " + str(accuracy_1a))
print("precision: " +str(precision_1a))
print("recall: " + str(recall_1a))
print("f1: " + str(f1_1a))
print("auc: " + str(auc_1a))
print(tn)
print(fp)
print(tp)
print(fn)

Answers for 1c: 
accuracy: 0.7372134038800705
precision: 0.0
recall: 0.0
f1: 0.0
auc: 0.5
418
0
0
149


/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
test_set_preds

In [ ]:
y_test

In [ ]:

# 01                    HarmVirtue
# 02                    HarmVice
# 03                    FairnessVirtue
# 04                    FairnessVice
# 05                    IngroupVirtue
# 06                    IngroupVice
# 07                    AuthorityVirtue
# 08                    AuthorityVice
# 09                    PurityVirtue
# 10                    PurityVice
# 11                    MoralityGeneral

Model 4


In [ ]:
d = {1:[], 2:[], 3:[], 4:[], 5:[],6:[],7:[],8:[],9:[],10:[],11:[] }


In [ ]:
d = {1:[], 2:[], 3:[], 4:[], 5:[],6:[],7:[],8:[],9:[],10:[],11:[] }

with open("MoralFoundations_dic_only.txt") as f:
    for line in f:
      #print(line)
      lst = line.split()
      if len(lst) >1:
        #print(lst)
        word = lst[0]

        word = word.replace('*', '')

        indecies = lst[1:]
        for i in indecies:
          toint = int(i)
          d[toint].append(word)

In [ ]:

X_train_1d = []
for text in train['request_text'].tolist():
  total_word_count = len(text.split())

  row =[]
  for i in range(len(d)):
    curr_set = d[i+1]
    set_word_count = 0
    for word in curr_set:
      if word in text:
        set_word_count +=1
    if total_word_count > 0:
      row.append(set_word_count/total_word_count)
    else:
      row.append(0)
  X_train_1d.append(row)

X_train_1d

In [ ]:

X_test_1d = []
for text in test['request_text'].tolist():
  total_word_count = len(text.split())

  row =[]
  for i in range(len(d)):
    curr_set = d[i+1]
    set_word_count = 0
    for word in curr_set:
      if word in text:
        set_word_count +=1
    if total_word_count > 0:
      row.append(set_word_count/total_word_count)
    else:
      row.append(0)
  X_test_1d.append(row)


In [ ]:
X_train_1d_np = np.array(X_train_1d)
X_train_1d_np.shape

(5104, 11)

In [ ]:
X_test_1d_np = np.array(X_test_1d)
X_test_1d_np.shape

(567, 11)

In [ ]:
y_train

In [ ]:
#svm4 = linear_model.SGDClassifier(loss='hinge', penalty='elasticnet')
#svm4.fit(X_train_1d_np, y_train)
from sklearn import svm
from sklearn.linear_model import SGDClassifier

#svm4 = SGDClassifier(loss="hinge", n_iter_no_change=10000).fit(X_train_1d_np, np.array(y_train))
svm4 = svm.SVC(kernel='linear')
svm4.fit(X_train_1d_np, np.array(y_train))

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [ ]:

import sklearn.metrics


test_set_preds = svm4.predict(X_test_1d_np)

accuracy_1a = svm4.score(X_test_1d_np, y_test)
precision_1a = sklearn.metrics.precision_score(y_test, test_set_preds)
recall_1a = sklearn.metrics.recall_score(y_test, test_set_preds)
f1_1a = sklearn.metrics.f1_score(y_test, test_set_preds)
auc_1a= sklearn.metrics.roc_auc_score(y_test, test_set_preds)

tn, fp, fn, tp = sklearn.metrics.confusion_matrix(y_test, test_set_preds).ravel()
specificity_1a = tn / (tn+fp)

print("Answers for 1d: ")
print("accuracy: " + str(accuracy_1a))
print("precision: " +str(precision_1a))
print("recall: " + str(recall_1a))
print("f1: " + str(f1_1a))
print("auc: " + str(auc_1a))
print(tn)
print(fp)
print(tp)
print(fn)

Answers for 1d: 
accuracy: 0.7372134038800705
precision: 0.0
recall: 0.0
f1: 0.0
auc: 0.5
418
0
0
149


/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
test_set_preds

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [ ]:
print(specificity_1a)

1.0
